man/man1/prte_info.1
.\" Automatically generated by Pandoc 2.14.2
.\"
.TH "prte_info" "1" "" "2021-08-23" "PRRTE"
.hy
.SH NAME
.PP
prte_info - Display information about the PRTE installation
.SH SYNOPSIS
.IP
.nf
\f[C]
prte-info [options]
\f[R]
.fi
.SH DESCRIPTION
.PP
\f[C]prte_info\f[R] provides detailed information about the PRTE
installation.
It can be useful for at least three common scenarios:
.IP "1." 3
Checking local configuration and seeing how PRTE was installed.
.IP "2." 3
Submitting bug reports / help requests to the PRTE community (see
https://github.com/openpmix/prte/issues.
.IP "3." 3
Seeing a list of installed PRTE plugins and querying what MCA parameters
they support.
.SH OPTIONS
.PP
\f[C]prte_info\f[R] accepts the following options:
.TP
\f[B]\f[CB]-a|--all\f[B]\f[R]
Show all configuration options and MCA parameters
.TP
\f[B]\f[CB]--arch\f[B]\f[R]
Show architecture PRTE was compiled on
.TP
\f[B]\f[CB]-c|--config\f[B]\f[R]
Show configuration options
.TP
\f[B]\f[CB]-gmca|--gmca <param> <value>\f[B]\f[R]
Pass global MCA parameters that are applicable to all contexts.
.TP
\f[B]\f[CB]-h|--help\f[B]\f[R]
Shows help / usage message
.TP
\f[B]\f[CB]--hostname\f[B]\f[R]
Show the hostname on which PRTE was configured and built.
.TP
\f[B]\f[CB]--internal\f[B]\f[R]
Show internal MCA parameters (not meant to be modified by users)
.TP
\f[B]\f[CB]-mca|--mca <param> <value>\f[B]\f[R]
Pass context-specific MCA parameters; they are considered global if
\f[C]--gmca\f[R] is not used and only one context is specified.
.TP
\f[B]\f[CB]--param <type> <component>\f[B]\f[R]
Show MCA parameters.
The first parameter is the type of the component to display; the second
parameter is the specific component to display (or the keyword
\f[C]all\f[R], meaning \[lq]display all components of this type\[rq]).
.TP
\f[B]\f[CB]--parsable\f[B]\f[R]
When used in conjunction with other parameters, the output is \[ga]
displayed in a machine-parsable format
.TP
\f[B]\f[CB]--parseable\f[B]\f[R]
Synonym for \f[C]--parsable\f[R].
.TP
\f[B]\f[CB]--path <type>\f[B]\f[R]
Show paths that PRTE was configured with.
Accepts the following parameters: \f[C]prefix\f[R], \f[C]bindir\f[R],
\f[C]libdir\f[R], \f[C]incdir\f[R], \f[C]pkglibdir\f[R],
\f[C]sysconfdir\f[R].
.TP
\f[B]\f[CB]--pretty\f[B]\f[R]
When used in conjunction with other parameters, the output is displayed
in `prettyprint' format (default).
.TP
\f[B]\f[CB]-v|--version <component> <scope>\f[B]\f[R]
Show version of PRTE or a component.
can be the keywords \f[C]ompi\f[R] or \f[C]all\f[R], the name of a
framework (e.g., \f[C]coll\f[R] shows all components in the
\f[C]coll\f[R] framework), or the name of a specific component (e.g.,
\f[C]pls:rsh\f[R] shows the information from the rsh PLS component).
can be one of: \f[C]full\f[R], \f[C]major\f[R], \f[C]minor\f[R],
\f[C]release\f[R], \f[C]greek\f[R], \f[C]git\f[R].
.SH EXAMPLES
.TP
\f[B]\f[CB]prte_info\f[B]\f[R]
Show the default output of options and listing of installed components
in a human-readable / prettyprint format.
.TP
\f[B]\f[CB]prte_info --parsable\f[B]\f[R]
Show the default output of options and listing of installed components
in a machine-parsable format.
.TP
\f[B]\f[CB]prte_info --param rmcast udp\f[B]\f[R]
Show the MCA parameters of the \[lq]udp\[rq] RMCAST component in a
human-readable / prettyprint format.
.TP
\f[B]\f[CB]prte_info --param rmcast udp --parsable\f[B]\f[R]
Show the MCA parameters of the \[lq]udp\[rq] RMCAST component in a
machine-parsable format.
.TP
\f[B]\f[CB]prte_info --path bindir\f[B]\f[R]
Show the \[lq]bindir\[rq] that PRTE was configured with.
.TP
\f[B]\f[CB]prte_info --version prte full --parsable\f[B]\f[R]
Show the full version numbers of PRTE (including the PRTE version
number) in a machine-readable format.
.TP
\f[B]\f[CB]prte_info --version rmcast major\f[B]\f[R]
Show the major version number of all RMCAST components in a prettyprint
format.
.TP
\f[B]\f[CB]prte_info --version rmcast:tcp minor\f[B]\f[R]
Show the minor version number of the TCP RMCAST component in a
prettyprint format.
.TP
\f[B]\f[CB]prte_info --all\f[B]\f[R]
Show \f[C]all\f[R] information about the PRTE installation, including
all components that can be found, the MCA parameters that they support,
versions of PRTE and the components, etc.
.SH AUTHORS
.PP
The PRTE maintainers \[en] see https://github.com/openpmix/prte or the
file \f[C]AUTHORS\f[R].
.PP
This manual page was originally contributed by Dirk Eddelbuettel
<edd@debian.org>, one of the Debian GNU/Linux maintainers for Open MPI,
and may be used by others.


NEWS
Copyright (c) 2004-2010 The Trustees of Indiana University and Indiana
                        University Research and Technology
                        Corporation.  All rights reserved.
Copyright (c) 2004-2006 The University of Tennessee and The University
                        of Tennessee Research Foundation.  All rights
                        reserved.
Copyright (c) 2004-2008 High Performance Computing Center Stuttgart,
                        University of Stuttgart.  All rights reserved.
Copyright (c) 2004-2006 The Regents of the University of California.
                        All rights reserved.
Copyright (c) 2006-2020 Cisco Systems, Inc.  All rights reserved.
Copyright (c) 2006      Voltaire, Inc. All rights reserved.
Copyright (c) 2006      Sun Microsystems, Inc.  All rights reserved.
                        Use is subject to license terms.
Copyright (c) 2006-2017 Los Alamos National Security, LLC.  All rights
                        reserved.
Copyright (c) 2010-2017 IBM Corporation.  All rights reserved.
Copyright (c) 2012      Oak Ridge National Labs.  All rights reserved.
Copyright (c) 2012      Sandia National Laboratories.  All rights reserved.
Copyright (c) 2012      University of Houston. All rights reserved.
Copyright (c) 2013      NVIDIA Corporation.  All rights reserved.
# Copyright (c) 2013-2019 Intel, Inc.  All rights reserved.
Copyright (c) 2018-2001 Amazon.com, Inc. or its affiliates.  All Rights
                        reserved.
Copyright (c) 2021      Nanook Consulting.  All rights reserved.
$COPYRIGHT$

Additional copyrights may follow

$HEADER$

===========================================================================

This file contains the main features as well as overviews of specific
bug fixes (and other actions) for each version of PRRTE since
version 1.0.

2.0.1: TBD
------------------------------------
+**** NOTE: As of v2.0.1, PRRTE no longer has a dependency on "pandoc"
+**** for building and installing man pages.
PR #1097: Update v2.0.1 for rc2
 - Update NEWS and VERSION
 - Circulate and report local completion status for collectives
 - Don't require pandoc to make a tarball
 - Do not register xml topos
 - Remove the Markdown source files for PRRTE man pages
PR #1089:
 - Cleanup timeout support to differentiate spawn vs execution time limits
PR #1086:
 - Use the info list array feature to construct registration data
 - Fix path check for pandoc
PR #1083:
 - Add MCA param to control device distance computation
 - Correctly report --output-directory and --output-filename as deprecated
 - Check for libevent minimum 2.0.21
 - Properly handle generic "orte" MCA params
 - Cleanups associated with IOF operations
 - Avoid use of MCA params for singleton and report-uri
 - autogen.pl: ignore all excluded components
 - Don't allow PRRTE IOF to block during delivery
 - Update ignores to include Coverity working directory
 - Ensure we hold the IOF data until the PMIx library is done with it
 - Add an "unlock" call to balance the code
 - Use correct ninfo when passing into APIs
 - Don't locally output stdout/err if prterun has a parent


2.0.0: 29 July 2021
------------------------------------
This is the initial production release of the PRRTE software system. It
contains full support of the PMIx v4.1 Standard plus extensions as provided
by OpenPMIx v4.1. It therefore requires support from that release or above.

Delineating all the changes from the original v1.0.0 tarball would be
overwhelming. The original tarball was provided solely as a means for
packagers to prepare for the eventual production version and was not
intended for general use. As expected, the code base has been extensively
revised/rewritten since that time.

Further changes shall be tracked here in follow-on releases.


1.0.0: 17 July, 2019
--------------------

Initial public release.


src/tools/prte_info/prte_info.c
/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil -*- */
/*
 * Copyright (c) 2004-2010 The Trustees of Indiana University and Indiana
 *                         University Research and Technology
 *                         Corporation.  All rights reserved.
 * Copyright (c) 2004-2006 The University of Tennessee and The University
 *                         of Tennessee Research Foundation.  All rights
 *                         reserved.
 * Copyright (c) 2004-2007 High Performance Computing Center Stuttgart,
 *                         University of Stuttgart.  All rights reserved.
 * Copyright (c) 2004-2005 The Regents of the University of California.
 *                         All rights reserved.
 * Copyright (c) 2007-2020 Cisco Systems, Inc.  All rights reserved
 * Copyright (c) 2010-2016 Los Alamos National Security, LLC.
 *                         All rights reserved.
 * Copyright (c) 2014-2020 Intel, Inc.  All rights reserved.
 * Copyright (c) 2015      Research Organization for Information Science
 *                         and Technology (RIST). All rights reserved.
 * Copyright (c) 2021      Nanook Consulting.  All rights reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 */

#include "prte_config.h"

#include <ctype.h>
#include <stdio.h>
#include <string.h>
#ifdef HAVE_UNISTD_H
#    include <unistd.h>
#endif
#ifdef HAVE_NETDB_H
#    include <netdb.h>
#endif
#ifdef HAVE_SYS_PARAM_H
#    include <sys/param.h>
#endif
#include <errno.h>
#include <signal.h>

#include "src/class/prte_object.h"
#include "src/class/prte_pointer_array.h"
#include "src/mca/base/base.h"
#include "src/mca/prteinstalldirs/prteinstalldirs.h"
#include "src/mca/schizo/base/base.h"
#include "src/prted/pmix/pmix_server.h"
#include "src/util/argv.h"
#include "src/util/cmd_line.h"
#include "src/util/error.h"
#include "src/util/path.h"
#include "src/util/proc_info.h"
#include "src/util/show_help.h"

#include "constants.h"
#include "src/include/frameworks.h"
#include "src/include/version.h"
#include "src/runtime/prte_locks.h"

#include "src/tools/prte_info/pinfo.h"

/*
 * Public variables
 */

bool prte_info_pretty = true;
prte_cmd_line_t *prte_info_cmd_line = NULL;

const char *prte_info_type_all = "all";
const char *prte_info_type_prte = "prte";
const char *prte_info_type_base = "base";

prte_pointer_array_t mca_types = {{0}};

prte_cmd_line_init_t info_cmd_line_opts[]
    = {{'h', "help", 0, PRTE_CMD_LINE_TYPE_BOOL, "This help message", PRTE_CMD_LINE_OTYPE_GENERAL},
       {'V', "version", 0, PRTE_CMD_LINE_TYPE_BOOL, "Print version and exit",
        PRTE_CMD_LINE_OTYPE_GENERAL},
       {'\0', "show-version", 2, PRTE_CMD_LINE_TYPE_STRING,
        "Show version of PRTE or a component.  The first parameter can be the keywords \"prte\" or "
        "\"all\", a framework name (indicating all components in a framework), or a "
        "framework:component string (indicating a specific component).  The second parameter can "
        "be one of: full, major, minor, release, greek, svn.",
        PRTE_CMD_LINE_OTYPE_GENERAL},
       {'\0', "param", 2, PRTE_CMD_LINE_TYPE_STRING,
        "Show MCA parameters.  The first parameter is the framework (or the keyword \"all\"); the "
        "second parameter is the specific component name (or the keyword \"all\").",
        PRTE_CMD_LINE_OTYPE_GENERAL},
       {'\0', "internal", 0, PRTE_CMD_LINE_TYPE_BOOL,
        "Show internal MCA parameters (not meant to be modified by users)",
        PRTE_CMD_LINE_OTYPE_GENERAL},
       {'\0', "path", 1, PRTE_CMD_LINE_TYPE_STRING,
        "Show paths that PRTE was configured with.  Accepts the following parameters: prefix, "
        "bindir, libdir, incdir, mandir, pkglibdir, sysconfdir",
        PRTE_CMD_LINE_OTYPE_GENERAL},
       {'\0', "arch", 0, PRTE_CMD_LINE_TYPE_BOOL, "Show architecture PRTE was compiled on",
        PRTE_CMD_LINE_OTYPE_GENERAL},
       {'c', "config", 0, PRTE_CMD_LINE_TYPE_BOOL, "Show configuration options",
        PRTE_CMD_LINE_OTYPE_GENERAL},
       {'\0', "hostname", 0, PRTE_CMD_LINE_TYPE_BOOL,
        "Show the hostname that PRTE was configured "
        "and built on",
        PRTE_CMD_LINE_OTYPE_GENERAL},
       {'a', "all", 0, PRTE_CMD_LINE_TYPE_BOOL, "Show all configuration options and MCA parameters",
        PRTE_CMD_LINE_OTYPE_GENERAL},

       /* End of list */
       {'\0', NULL, 0, PRTE_CMD_LINE_TYPE_NULL, NULL}};

int main(int argc, char *argv[])
{
    int ret = 0;
    bool want_help = false;
    bool cmd_error = false;
    bool acted = false;
    bool want_all = false;
    int i;
    char *str;

    /* protect against problems if someone passes us thru a pipe
     * and then abnormally terminates the pipe early */
    signal(SIGPIPE, SIG_IGN);

    /* Initialize the argv parsing stuff */
    if (PRTE_SUCCESS != (ret = prte_init_util(PRTE_PROC_MASTER))) {
        prte_show_help("help-prte-info.txt", "lib-call-fail", true, "prte_init_util", __FILE__,
                       __LINE__, NULL);
        exit(ret);
    }

    prte_info_cmd_line = PRTE_NEW(prte_cmd_line_t);
    if (NULL == prte_info_cmd_line) {
        ret = errno;
        prte_show_help("help-prte-info.txt", "lib-call-fail", true, "prte_cmd_line_create",
                       __FILE__, __LINE__, NULL);
        exit(ret);
    }

    ret = prte_cmd_line_add(prte_info_cmd_line, info_cmd_line_opts);
    if (PRTE_SUCCESS != ret) {
        return ret;
    }

    /* Do the parsing */
    ret = prte_cmd_line_parse(prte_info_cmd_line, false, false, argc, argv);
    if (PRTE_SUCCESS != ret) {
        if (PRTE_ERR_SILENT != ret) {
            fprintf(stderr, "%s: command line error (%s)\n", argv[0], prte_strerror(ret));
        }
        cmd_error = true;
    }
    if (!cmd_error
        && (prte_cmd_line_is_taken(prte_info_cmd_line, "help")
            || prte_cmd_line_is_taken(prte_info_cmd_line, "h"))) {
        char *usage;

        want_help = true;
        usage = prte_cmd_line_get_usage_msg(prte_info_cmd_line, false);
        str = prte_show_help_string("help-prte-info.txt", "usage", true, usage);
        if (NULL != str) {
            printf("%s", str);
            free(str);
        }
        free(usage);
    }
    if (cmd_error || want_help) {
        prte_mca_base_close();
        PRTE_RELEASE(prte_info_cmd_line);
        exit(cmd_error ? 1 : 0);
    }

    if (prte_cmd_line_is_taken(prte_info_cmd_line, "version")) {
        fprintf(stdout, "PRTE v%s\n\n%s\n", PRTE_VERSION, PACKAGE_BUGREPORT);
        exit(0);
    }

    /* setup the mca_types array */
    PRTE_CONSTRUCT(&mca_types, prte_pointer_array_t);
    prte_pointer_array_init(&mca_types, 256, INT_MAX, 128);

    /* add a type for prte itself */
    prte_pointer_array_add(&mca_types, "prte");

    /* add a type for hwloc */
    prte_pointer_array_add(&mca_types, "hwloc");

    /* let the pmix server register params */
    pmix_server_register_params();
    /* add those in */
    prte_pointer_array_add(&mca_types, "pmix");

    /* push all the types found by autogen */
    for (i = 0; NULL != prte_frameworks[i]; i++) {
        prte_pointer_array_add(&mca_types, prte_frameworks[i]->framework_name);
    }

    /* Execute the desired action(s) */

    if (prte_cmd_line_is_taken(prte_info_cmd_line, "prte_info_pretty")) {
        prte_info_pretty = true;
    } else if (prte_cmd_line_is_taken(prte_info_cmd_line, "parsable")
               || prte_cmd_line_is_taken(prte_info_cmd_line, "parseable")) {
        prte_info_pretty = false;
    }

    want_all = prte_cmd_line_is_taken(prte_info_cmd_line, "all");
    if (want_all || prte_cmd_line_is_taken(prte_info_cmd_line, "show-version")) {
        prte_info_do_version(want_all, prte_info_cmd_line);
        acted = true;
    }
    if (want_all || prte_cmd_line_is_taken(prte_info_cmd_line, "path")) {
        prte_info_do_path(want_all, prte_info_cmd_line);
        acted = true;
    }
    if (want_all || prte_cmd_line_is_taken(prte_info_cmd_line, "arch")) {
        prte_info_do_arch();
        acted = true;
    }
    if (want_all || prte_cmd_line_is_taken(prte_info_cmd_line, "hostname")) {
        prte_info_do_hostname();
        acted = true;
    }
    if (want_all || prte_cmd_line_is_taken(prte_info_cmd_line, "config")) {
        prte_info_do_config(true);
        acted = true;
    }
    if (want_all || prte_cmd_line_is_taken(prte_info_cmd_line, "param")) {
        prte_info_do_params(want_all, prte_cmd_line_is_taken(prte_info_cmd_line, "internal"));
        acted = true;
    }

    /* If no command line args are specified, show default set */

    if (!acted) {
        prte_info_show_prte_version(prte_info_ver_full);
        prte_info_show_path(prte_info_path_prefix, prte_install_dirs.prefix);
        prte_info_do_arch();
        prte_info_do_hostname();
        prte_info_do_config(false);
        prte_info_components_open();
        for (i = 0; i < mca_types.size; ++i) {
            if (NULL == (str = (char *) prte_pointer_array_get_item(&mca_types, i))) {
                continue;
            }
            prte_info_show_component_version(str, prte_info_component_all, prte_info_ver_full,
                                             prte_info_type_all);
        }
    }

    /* All done */
    prte_info_components_close();
    PRTE_DESTRUCT(&mca_types);
    prte_mca_base_close();

    return 0;
}
